File: src/sign_mag_truth_table_generator.cpp
#include <iostream>
#include <bitset>
#include <cmath>
#include <algorithm>
#include <fstream>

using namespace std;

class Calc
{
public:
    static const uint32_t A_SGN_8_BIT_MASK = 0b00001000;
    static const uint32_t A_MAG_8_BIT_MASK = 0b00000111;
    static const uint32_t B_SGN_8_BIT_MASK = 0b10000000;
    static const uint32_t B_MAG_8_BIT_MASK = 0b01110000;
    static const uint32_t R_MAG_8_BIT_MASK = 0b00000111;

    static const uint32_t A_SGN_16_BIT_MASK = 0b0000000010000000;
    static const uint32_t A_MAG_16_BIT_MASK = 0b0000000001111111;
    static const uint32_t B_SGN_16_BIT_MASK = 0b1000000000000000;
    static const uint32_t B_MAG_16_BIT_MASK = 0b0111111100000000;
    static const uint32_t R_MAG_16_BIT_MASK = 0b0000000001111111;

    int addr_size;
    int data_size;
    uint32_t depth;
    ofstream truth_table_out;
    Calc(int addr_size)
    {
        this->addr_size = addr_size;
        if (this->addr_size != 8 && this->addr_size != 16)
        {
            throw std::invalid_argument("Only supports 8 and 16 bit addresses");
        }
        this->data_size = (int)(addr_size / 2);
        this->depth = pow(2, addr_size);
        this->calculate();
    }

    uint32_t get_a_sgn(uint32_t val)
    {
        return (val & (this->addr_size == 8 ? A_SGN_8_BIT_MASK : A_SGN_16_BIT_MASK)) >> (this->data_size - 1);
    }

    uint32_t get_a_mag(uint32_t val)
    {
        return val & (this->addr_size == 8 ? A_MAG_8_BIT_MASK : A_MAG_16_BIT_MASK);
    }

    uint32_t get_b_sgn(uint32_t val)
    {
        return (val & (this->addr_size == 8 ? B_SGN_8_BIT_MASK : B_SGN_16_BIT_MASK)) >> ((data_size * 2) - 1);
    }

    uint32_t get_b_mag(uint32_t val)
    {
        return (val & (this->addr_size == 8 ? B_MAG_8_BIT_MASK : B_MAG_16_BIT_MASK)) >> (data_size);
    }

    uint32_t get_r_mag(uint32_t val)
    {
        return val & (this->addr_size == 8 ? R_MAG_8_BIT_MASK : R_MAG_16_BIT_MASK);
    }

    void calculate()
    {
        string file_name = "rom_" + to_string(this->addr_size) + "_sign_mag.txt";
        truth_table_out.open(file_name);

        for (uint32_t i = 0; i < this->depth; i++)
        {
            uint32_t a_sgn = this->get_a_sgn(i);
            uint32_t a_mag = this->get_a_mag(i);
            uint32_t b_sgn = this->get_b_sgn(i);
            uint32_t b_mag = this->get_b_mag(i);

            uint32_t r_mag;
            uint32_t r_sgn;

            // If the two operands have the same sign, add the magnitude and keep the sign
            if (a_sgn == b_sgn)
            {
                r_mag = a_mag + b_mag;
                if (r_mag == 0)
                {
                    r_sgn = 0;
                }
                else
                {
                    r_sgn = a_sgn;
                }
            }
            else
            {
                // If the two operands have different signs, subtract the smaller magnitude from the larger one
                r_mag = this->get_r_mag((max(a_mag, b_mag) - min(a_mag, b_mag)));
                // And keep the larger one's sign
                if (a_mag > b_mag)
                {
                    r_sgn = a_sgn;
                }
                else if (b_mag > a_mag)
                {
                    r_sgn = b_sgn;
                }
                else
                {
                    r_sgn = 0;
                }
            }

            if (this->addr_size == 8)
            {
                truth_table_out << bitset<1>(r_sgn)
                                << bitset<3>(r_mag)
                                << endl;
            }
            else
            {
                truth_table_out << bitset<1>(r_sgn)
                                << bitset<7>(r_mag)
                                << endl;
            }
        }

        truth_table_out.close();
    }
};

int main()
{
    Calc calc_8(8);
    Calc calc_16(16);
    return 0;
}

File: src/sign_mag_truth_table_generator.py
import itertools

def create_sign_mag_truth_table(addr_size=8, verbose=True):
    ADDR_SIZE = addr_size
    FILE=f("rom1_{ADDR_SIZE}_bit_sign_mag.txt");

    MID = int(ADDR_SIZE/2)
    DATA_SIZE = MID
    VERBOSE = verbose

    # Generate 2^ADDR_SIZE truth table
    truth_table_in = [i for i in itertools.product([0,1], repeat=ADDR_SIZE)]

    # Convert int to string binary with at least n bits
    to_bin = lambda x, n: format(x, 'b').zfill(n)

    # Variable to hold output of truth table
    truth_table_out = open(FILE, 'w')

    # iterate the truth table
    for i, addr in enumerate(truth_table_in):

        a_sgn = addr[0]
        a_mag = ''.join(map(str,addr[1:MID]))
        b_sgn = addr[MID]
        b_mag = ''.join(map(str,addr[MID+1:ADDR_SIZE]))

        a_int = int(a_mag, 2)
        b_int = int(b_mag, 2)

        # If the two operands have the same sign, add the magnitude and keep the sign
        if (a_sgn == b_sgn):
            r_int = a_int + b_int
            # If the result magnitude is 0 just use +0 
            if r_int == 0:
                r_sgn = 0
            else:
                r_sgn = a_sgn
        else:
            # If the two operands have different signs, subtract the smaller magnitude from the larger one
            r_int = max(a_int, b_int) - min(a_int, b_int)
            # And keep the larger one's sign
            if a_int > b_int:
                r_sgn = a_sgn
            elif b_int > a_int:
                r_sgn = b_sgn
            else:
                r_sgn = 0

        r_bin = to_bin(r_int, DATA_SIZE-1)

        if VERBOSE:
            log_msg = f("{i+1}: {addr} = {a_sgn} {a_mag} + {b_sgn} {b_mag} = {r_sgn} {r_bin}");
            if len(r_bin) > (DATA_SIZE -1):
                log_msg += " overflow"
            print(log_msg)

        # Save result to the file
        if len(r_bin) > DATA_SIZE - 1:
            r_bin_str = str(r_bin)[1:DATA_SIZE]
        else:
            r_bin_str = str(r_bin)

        truth_table_out.write(f('{r_sgn}{r_bin_str}\n'))

    truth_table_out.close()
    
if __name__=="__main__":
    create_sign_mag_truth_table(8)
    create_sign_mag_truth_table(16)
